File: backbone-go/examples/clean-api-go/domain/entities/product.go
// Package entities contains domain entities
package entities

import (
	"time"

	"github.com/google/uuid"
)

// Product represents a product entity in the domain
type Product struct {
	ID          string
	Name        string
	Description string
	Price       float64
	Category    string
	Stock       int
	Active      bool
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// NewProduct creates a new product with validation
func NewProduct(name, description, category string, price float64, stock int) (*Product, error) {
	product := &Product{
		ID:          uuid.New().String(),
		Name:        name,
		Description: description,
		Price:       price,
		Category:    category,
		Stock:       stock,
		Active:      true,
		CreatedAt:   time.Now(),
		UpdatedAt:   time.Now(),
	}

	if err := product.Validate(); err != nil {
		return nil, err
	}

	return product, nil
}

// Validate validates the product
func (p *Product) Validate() error {
	if p.Name == "" || len(p.Name) < 3 {
		return &ValidationError{
			Field:   "name",
			Message: "Product name must be at least 3 characters",
			Code:    11001001,
		}
	}

	if p.Price <= 0 {
		return &ValidationError{
			Field:   "price",
			Message: "Product price must be greater than 0",
			Code:    11001002,
		}
	}

	if p.Stock < 0 {
		return &ValidationError{
			Field:   "stock",
			Message: "Product stock cannot be negative",
			Code:    11001003,
		}
	}

	if p.Category == "" {
		return &ValidationError{
			Field:   "category",
			Message: "Product category is required",
			Code:    11001004,
		}
	}

	return nil
}

// UpdateStock updates the stock quantity
func (p *Product) UpdateStock(quantity int) error {
	if quantity < 0 {
		return &ValidationError{
			Field:   "stock",
			Message: "Stock quantity cannot be negative",
			Code:    11001003,
		}
	}

	p.Stock = quantity
	p.UpdatedAt = time.Now()
	return nil
}

// UpdatePrice updates the price
func (p *Product) UpdatePrice(price float64) error {
	if price <= 0 {
		return &ValidationError{
			Field:   "price",
			Message: "Product price must be greater than 0",
			Code:    11001002,
		}
	}

	p.Price = price
	p.UpdatedAt = time.Now()
	return nil
}

// Deactivate marks the product as inactive
func (p *Product) Deactivate() {
	p.Active = false
	p.UpdatedAt = time.Now()
}

// Activate marks the product as active
func (p *Product) Activate() {
	p.Active = true
	p.UpdatedAt = time.Now()
}

// IsInStock checks if product is in stock
func (p *Product) IsInStock() bool {
	return p.Stock > 0
}

// ValidationError represents a domain validation error
type ValidationError struct {
	Field   string
	Message string
	Code    int
}

// Error implements the error interface
func (e *ValidationError) Error() string {
	return e.Message
}

File: backbone-python/examples/clean_api_python/domain/entities/product.py
"""Product entity - Domain layer"""
from datetime import datetime
from uuid import uuid4
from typing import Optional


class ValidationError(Exception):
    """Domain validation error"""
    def __init__(self, field: str, message: str, code: int):
        self.field = field
        self.message = message
        self.code = code
        super().__init__(message)


class Product:
    """Product entity in the domain"""
    
    def __init__(
        self,
        name: str,
        description: str,
        category: str,
        price: float,
        stock: int,
        id: Optional[str] = None,
        active: bool = True,
        created_at: Optional[datetime] = None,
        updated_at: Optional[datetime] = None
    ):
        self.id = id or str(uuid4())
        self.name = name
        self.description = description
        self.price = price
        self.category = category
        self.stock = stock
        self.active = active
        self.created_at = created_at or datetime.now()
        self.updated_at = updated_at or datetime.now()
        
        self.validate()
    
    def validate(self) -> None:
        """Validates the product"""
        if not self.name or len(self.name) < 3:
            raise ValidationError(
                "name",
                "Product name must be at least 3 characters",
                11001001
            )
        
        if self.price <= 0:
            raise ValidationError(
                "price",
                "Product price must be greater than 0",
                11001002
            )
        
        if self.stock < 0:
            raise ValidationError(
                "stock",
                "Product stock cannot be negative",
                11001003
            )
        
        if not self.category:
            raise ValidationError(
                "category",
                "Product category is required",
                11001004
            )
    
    def update_stock(self, quantity: int) -> None:
        """Updates the stock quantity"""
        if quantity < 0:
            raise ValidationError(
                "stock",
                "Stock quantity cannot be negative",
                11001003
            )
        self.stock = quantity
        self.updated_at = datetime.now()
    
    def update_price(self, price: float) -> None:
        """Updates the price"""
        if price <= 0:
            raise ValidationError(
                "price",
                "Product price must be greater than 0",
                11001002
            )
        self.price = price
        self.updated_at = datetime.now()
    
    def deactivate(self) -> None:
        """Marks the product as inactive"""
        self.active = False
        self.updated_at = datetime.now()
    
    def activate(self) -> None:
        """Marks the product as active"""
        self.active = True
        self.updated_at = datetime.now()
    
    def is_in_stock(self) -> bool:
        """Checks if product is in stock"""
        return self.stock > 0
    
    def to_dict(self) -> dict:
        """Converts to dictionary"""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "price": self.price,
            "category": self.category,
            "stock": self.stock,
            "active": self.active,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat()
        }
